hiro/cocoa/widget/hex-edit.cpp

```diff
@@ -5,43 +5,157 @@
 -(id) initWith:(hiro::mHexEdit&)hexEditReference {
   if(self = [super initWithFrame:NSMakeRect(0, 0, 0, 0)]) {
     hexEdit = &hexEditReference;
+    if (tableView = [[NSTableView alloc] initWithFrame:[self bounds]]) {
+      [tableView setDataSource:self];
+      [tableView setHeaderView:nil];
+      
+      // remove padding and stuff to fit all 16 columns
+      [tableView setIntercellSpacing:NSMakeSize(0, 0)];
+      [tableView setStyle:NSTableViewStylePlain];
+      NSTableColumn* addressCol = [[NSTableColumn alloc] initWithIdentifier:@"Address"];
+      addressCol.editable = NO;
+      [tableView addTableColumn:addressCol];
+      
+      for (int i = 0; i < hexEdit->columns(); i++) {
+        NSTableColumn* col = [[NSTableColumn alloc] initWithIdentifier:[NSString stringWithFormat:@"%d",
+                                                                        i]];
+        col.width = 20;
+        col.title = @"";
+        [tableView addTableColumn:col];
+      }
+      /*
+       NSTableColumn* hexCol = [[NSTableColumn alloc] initWithIdentifier:@"Hex"];
+       [hexCol setTitle:@"Hex"];
+       [tableView addTableColumn:hexCol];
+       */
+      
+      NSTableColumn* ansiCol = [[NSTableColumn alloc] initWithIdentifier:@"ANSI"];
+      ansiCol.editable = NO;
+      [tableView addTableColumn:ansiCol];
+      
+      // must programmatically do this so the table shows up.
+      [self setDocumentView:tableView];
+      [self setHasVerticalScroller:YES];
+    }
   }
   return self;
 }
 
-@end
-
-namespace hiro {
-
-auto pHexEdit::construct() -> void {
-  cocoaView = cocoaHexEdit = [[CocoaHexEdit alloc] initWith:self()];
+- (NSTableView *) tableView {
+  return tableView;
 }
 
-auto pHexEdit::destruct() -> void {
-  [cocoaView removeFromSuperview];
+- (NSInteger) numberOfRowsInTableView:(NSTableView *)tableView {
+  return hexEdit->rows();
```

Comment:
Hmm; it looks like in other hex-edit implementations, `rows` is being dynamically set based on the amount of memory being looked at (`hexEdit->length()`. For Cocoa here, we aren't currently doing that, so this value appears to always remain at 16. I think we need to derive the number of rows from `hexEdit->length` in this method, or else set `hexEdit`'s `rows` somewhere else based on the total memory length.

Reply:
Thank you – I was misunderstanding the functionality. Newest commit fixes this so rows = length / columns and properly implements address goto!